main.go

```diff
@@ -132,6 +171,60 @@ func main() {
 	}
 }
 
+func deleteAction(ctx context.Context, clientSet client.Client, hr *fluxhelmv2beta1.HelmRelease, timeout, interval time.Duration) {
+	instance := &fluxhelmv2beta1.HelmRelease{}
+	key := client.ObjectKey{
+		Name:      hr.Name,
+		Namespace: hr.Namespace,
+	}
+	if err := clientSet.Get(ctx, key, instance); client.IgnoreNotFound(err) != nil {
+		log.Fatalf("Failed to get instance of the helmrelease with %v", err)
+	} else if err != nil {
+		// This means that the object was not found
+		log.Errorf("Did not find the helm release: %v", hr.Name)
+	} else {
+		instance.Annotations = hr.Annotations
+		instance.Labels = hr.Labels
+		instance.Spec = hr.Spec
+
+		log.Infof("Found the helm release: %v, deleting the release...", hr.Name)
+		// Delete the HelmRelease
+		retry(ctx, func() error { return clientSet.Delete(ctx, instance) }, interval)
+	}
+
+	if err := PollHelmReleaseDeleted(ctx, clientSet, hr, timeout); err != nil {
+		log.Errorf("Failed to poll the deletion of the helm release with %v", err)
+		if err := clientSet.Get(ctx, key, instance); client.IgnoreNotFound(err) != nil {
+			log.Fatalf("Failed to get instance of the helmrelease with %v", err)
+		} else if err != nil {
+			log.Errorf("Did not find the helm release: %v, the object is deleted", hr.Name)
```

Comment:
Can we just return here if we end up in this block. In that case the following else block can be removed. The code in that block can be refactored a bit :
```golang
func cleanupHelmRelease(hr HelmRelease, ...other args) error {
    	return PollHelmReleaseDeleted(ctx, clientSet, hr, timeout)
}

func forceCleanupHelmRelease(hr HelmRelease, ...other args) error {
    	if err := PollHelmReleaseDeleted(ctx, clientSet, hr, timeout); err != nil {
        // Remove finalizer from HelmRelease
        // `helm uninstall`
        return
      }
}
// First cleanup call
if err := deleteAction(cleanupHelmRelease(hr, ...)); if err != nil {
  return deleteAction(forceCleanupHelmRelease(hr, ...))
}

func deleteAction(cleanupFn func(hr HelmRelease, ...other args) error) error {
  // GET
  // if FOUND:
  if err := cleanupFunc(); err != nil {
    return deleteAction(forceCleanupFunc())
  }
}
```

Reply:
I refactored the code a little bit.